core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java

```diff
@@ -353,15 +320,22 @@ private void growPointerArrayIfNecessary() throws IOException {
     assert(inMemSorter != null);
     if (!inMemSorter.hasSpaceForAnotherRecord()) {
       logger.debug("Attempting to expand sort pointer array");
-      final long oldPointerArrayMemoryUsage = inMemSorter.getMemoryUsage();
-      final long memoryToGrowPointerArray = oldPointerArrayMemoryUsage * 2;
-      final long memoryAcquired = taskMemoryManager.acquireExecutionMemory(memoryToGrowPointerArray);
-      if (memoryAcquired < memoryToGrowPointerArray) {
-        taskMemoryManager.releaseExecutionMemory(memoryAcquired);
-        spill();
-      } else {
-        inMemSorter.expandPointerArray();
-        taskMemoryManager.releaseExecutionMemory(oldPointerArrayMemoryUsage);
+      long used = inMemSorter.getMemoryUsage();
+      long needed = inMemSorter.getMemoryToExpand();
+      try {
+        acquireMemory(used + needed);  // could trigger spilling
+        if (inMemSorter.hasSpaceForAnotherRecord()) {
+          releaseMemory(used + needed);
+        } else {
+          logger.debug("Expand sort pointer array");
+          inMemSorter.expandPointerArray();
+          releaseMemory(used);
+        }
+      } catch (OutOfMemoryError oom) {
```

Comment:
We probably shouldn't catch `OutOfMemoryError`, since this could lead to problems in case we catch an OOM thrown by the JVM and not by our own code. Instead, we might want to throw our own `SparkOutOfMemoryError` exception or something like that.